### PR TITLE
Disallow non-element params to service getters

### PR DESCRIPTION
**Pull request by choumx**
Partial for #16322. Goal is to prevent bugs like #16043 in the future.

- Fix a few instances of `window.document` being passed to service getters.
- Change all service getter methods to accept `!Element` instead of `!Node` to prevent future bugs.
- Create `getServiceForDocDeprecated(nodeOrDoc, id)` to keep a few remaining usages (startup services) for follow-up fix.

/to @jridgewell @danielrozenberg /cc @jpettitt @aghassemi

### CODE REVIEW

File: src/element-service.js [hunk 1]
```diff
@@ -139,22 +140,22 @@ export function getElementServiceIfAvailableForDoc(
...
+  const s = getExistingServiceForDocInEmbedScope(elementOrAmpDoc, id);
   if (s) {
     return /** @type {!Promise<?Object>} */ (Promise.resolve(s));
   }
   // Return embed-scope element service promise if scheduled.
-  if (nodeOrDoc.nodeType) {
+  if (elementOrAmpDoc.nodeType) {
     const win = toWin(/** @type {!Document} */ (
-      nodeOrDoc.ownerDocument || nodeOrDoc).defaultView);
+      elementOrAmpDoc.ownerDocument || elementOrAmpDoc).defaultView);
```
[the remark this answers is not in the document]
Comment (choumx) [reply src/element-service.js]: Done.

File: src/service.js [hunk 2]
```diff
@@ -101,26 +101,26 @@ export function getExistingServiceInEmbedScope(win, id, opt_fallbackToTopWin) {
...
 export function getExistingServiceForDocInEmbedScope(
-  nodeOrDoc, id, opt_fallbackToTopWin) {
+  elementOrAmpDoc, id, opt_fallbackToTopWin) {
   // First, try to resolve via local (embed) window.
-  if (nodeOrDoc.nodeType) {
+  if (elementOrAmpDoc.nodeType) {
     // If a node is passed, try to resolve via this node.
     const win = toWin(/** @type {!Document} */ (
-      nodeOrDoc.ownerDocument || nodeOrDoc).defaultView);
+      elementOrAmpDoc.ownerDocument || elementOrAmpDoc).defaultView);
```
[the remark this answers is not in the document]
Comment (choumx) [reply src/service.js]: Oh man, nice catch. Thanks.

File: src/service/navigation.js [hunk 3]
```diff
@@ -436,7 +441,7 @@ export class Navigation {
    * @private
    */
   parseUrl_(url) {
-    return Services.urlForDoc(this.rootNode_).parse(url);
+    return Services.urlForDoc(this.ampdoc).parse(url);
```
[the remark this answers is not in the document]
Comment (choumx) [reply src/service/navigation.js]: I considered that, but I don't think local scope is important for URL parsing. Line 168 already grabs the URL service from `this.ampdoc`.

**Comment by choumx** [comment]
> Why can't we just have it accept a Document?

It would only work for single docs since shadow AMPs don't have a Document, and I want to prevent misuse. Part of the problem is Document vs. AmpDoc conflation e.g. `fooForDoc` is really misleading.

File: ads/google/a4a/utils.js [hunk 4]
```diff
@@ -429,14 +429,14 @@ function elapsedTimeWithCeiling(time, start) {
 /**
  * `nodeOrDoc` must be passed for correct behavior in shadow AMP (PWA) case.
  * @param {!Window} win
- * @param {!Node|!../../../src/service/ampdoc-impl.AmpDoc} nodeOrDoc
```
Comment (danielrozenberg) [inline ads/google/a4a/utils.js]: `!../../../src/service/ampdoc-impl.AmpDoc` makes me :cry:

File: src/service/navigation.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jridgewell) [reply src/service/navigation.js]: For FIEs, it's critical. FIE have a `<base>` element that changes where their relative URLs resolve from.

[the remark this answers is not in the document]
Comment (choumx) [reply src/service/navigation.js]: TIL, good call. I think I was too aggressive here -- reverted APIs for FIE-scope services.

**Comment by choumx** [comment]
@jridgewell Friendly ping.

File: src/service.js [hunk 5]
```diff
@@ -101,25 +101,29 @@ export function getExistingServiceInEmbedScope(win, id, opt_fallbackToTopWin) {
 /**
  * Returns a service with the given id. Assumes that it has been constructed
  * already.
- * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+ *
+ * Unlike most service getters, passing `Node` is necessary for some FIE-scope
+ * services since sometimes we only have the FIE Document for context.
+ *
+ * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrAmpDoc
```
Comment (jridgewell) [inline src/service.js]: Could we change this to `!Element|!ShadowRoot|!AmpDoc`? That we we fix the core bug here (passing in a `Document`) but still allow `ShadowRoot`s?

File: src/service.js [hunk 6]
```diff
@@ -101,6 +101,10 @@ export function getExistingServiceInEmbedScope(win, id, opt_fallbackToTopWin) {
 /**
  * Returns a service with the given id. Assumes that it has been constructed
  * already.
+ *
+ * Unlike most service getters, passing `Node` is necessary for some FIE-scope
+ * services since sometimes we only have the FIE Document for context.
+ *
  * @param {!Node|!./service/ampdoc-impl.AmpDoc} nodeOrDoc
```
Comment (jridgewell) [inline src/service.js]: Could we change this to `!Element|!ShadowRoot|!AmpDoc`? That we we fix the core bug here (passing in a `Document`) but still allow `ShadowRoot`s?

**Pull request merged**